app/models/course/experience_points_disbursement.rb

```diff
@@ -0,0 +1,104 @@
+  #
+  # @param [Course] course The course the disbursement is for
+  # @return [Course] The course the disbursement is for
+  def course=(course)
+    @course_users = course.course_users.students.with_approved_state.
+                    includes(:experience_points_records)
+    @experience_points_records = @course_users.map do |course_user|
+      course_user.experience_points_records.build
+    end
+    course
+  end
+
+  # Processes course user attributes and builds experience points records with the given
+  # attributes for those course user attributes hashes that represent a valid experience points
+  # award.
+  #
+  # @param [Hash] attributes Course users attributes
+  # @return [Hash] Course users attributes
+  def course_users_attributes=(attributes)
+    course_user_hashes = filter_valid_course_user_hashes(attributes.values)
+    load_course_users(course_user_hashes)
+
+    @experience_points_records = course_user_hashes.map do |course_user_hash|
+      points_awarded = course_user_hash[:experience_points_record][:points_awarded]
+      awardee = find_course_user(course_user_hash[:id].to_i)
+      awardee && awardee.experience_points_records.build(points_awarded: points_awarded,
+                                                         reason: reason)
+    end.compact
+    attributes
+  end
+
+  # Saves each newly built experience points record so that points are disbursed to students.
+  # Saving is done as a single transaction firstly for better performance, and secondly
+  # since if some of them fail, there is no easy way for the disburser to discover which students
+  # the points were not successfully awarded for.
+  #
+  # @return [Boolean] True if bulk saving was successful.
+  def save
+    outcome = Course::ExperiencePointsRecord.transaction do
+      @experience_points_records.map(&:save!)
+    end
+  rescue => e
+    errors.add(:experience_points_record, e.record.errors.full_messages.to_sentence)
+    false
+  else
+    @recipient_count = outcome.length
+    true
+  end
+
+  private
+
+  # Filters course users hashes that represents a valid experience points award.
+  #
+  # @param [Array<Hash>] hashes Course user attributes hashes
+  # @return [Array<Hash>] Valid course user attributes hashes
+  def filter_valid_course_user_hashes(hashes)
+    hashes.reject do |course_user_hash|
+      course_user_hash[:id].nil? ||
+        course_user_hash[:experience_points_record].nil? ||
+        course_user_hash[:experience_points_record][:points_awarded].blank? ||
+        course_user_hash[:experience_points_record][:points_awarded] == '0'
+    end
+  end
+
+  # Loads the associated +CourseUser+ for each of the given course user hash.
+  #
+  # @param [Array<Hash>] course_user_hashes Course user attributes hashes
+  def load_course_users(course_user_hashes)
+    course_users_ids = course_user_hashes.map { |course_user_hash| course_user_hash[:id] }
+    @course_users = CourseUser.includes(:course).find(course_users_ids)
```

Comment:
I eager loaded to avoid an N+1 query when the records are authorized. `course` has not been included yet?